src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs

```diff
@@ -249,24 +282,81 @@ private CSharpAttributeData GetAttribute(BoundAttribute boundAttribute, BindingD
+                    (AttributeSyntax)boundAttribute.Syntax,
+                    argsToParamsOpt,
+                    diagnostics,
+                    boundAttribute.ConstructorExpanded,
+                    ref hasErrors);
             }
 
             CompoundUseSiteInfo<AssemblySymbol> useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
             bool isConditionallyOmitted = IsAttributeConditionallyOmitted(attributeType, boundAttribute.SyntaxTree, ref useSiteInfo);
             diagnostics.Add(boundAttribute.Syntax, useSiteInfo);
 
-            return new SourceAttributeData(boundAttribute.Syntax.GetReference(), attributeType, attributeConstructor, constructorArguments, constructorArgumentsSourceIndices, namedArguments, hasErrors, isConditionallyOmitted);
+            return new SourceAttributeData(
+                boundAttribute.Syntax.GetReference(),
+                attributeType,
+                attributeConstructor,
+                rewrittenArguments,
+                makeSourceIndices(),
+                namedArguments,
+                hasErrors,
+                isConditionallyOmitted);
+
+            ImmutableArray<int> makeSourceIndices()
+            {
+                Debug.Assert(attributeConstructor.ParameterCount == rewrittenArguments.Length || hasErrors);
+                var lengthAfterRewriting = rewrittenArguments.Length;
+                if (lengthAfterRewriting == 0)
+                {
+                    return default;
+                }
+
+                // make source indices if we have anything that doesn't map 1:1 from arguments to parameters:
+                // 1. implicit default arguments
+                // 2. reordered labeled arguments
+                // 3. expanded params calls
+                var defaultArguments = boundAttribute.ConstructorDefaultArguments;
+                if (argsToParamsOpt.IsDefault && !boundAttribute.ConstructorExpanded)
+                {
+                    var hasDefaultArgument = false;
+                    var lengthBeforeRewriting = arguments.Length;
+                    for (var i = 0; i < lengthBeforeRewriting; i++)
+                    {
+                        if (defaultArguments[i])
+                        {
+                            hasDefaultArgument = true;
+                            break;
+                        }
+                    }
+                    if (!hasDefaultArgument)
+                    {
+                        return default;
+                    }
+                }
+
+                var constructorArgumentSourceIndices = ImmutableArray.CreateBuilder<int>(initialCapacity: lengthAfterRewriting);
+                for (int i = 0; i < lengthAfterRewriting; i++)
+                {
+                    constructorArgumentSourceIndices.Add(
+                        defaultArguments[i] ? -1 :
+                        argsToParamsOpt.IsDefault ? i :
+                        argsToParamsOpt[i]);
```

Comment:
>argsToParamsOpt[i] [](http://example.com/codeflow?start=24&length=18)

Is this correct? It looks like `constructorArgumentSourceIndices[i]` is index of argument corresponding to parameter `i`, but `argsToParamsOpt[i]` is the index of the parameter corresponding to argument `i`.

Reply:
I think you're right about this. What might be needed is more like `argsToParamsOpt.IndexOf(1)`, maybe an assertion that the resulting index is non-negative. We also should add a test where this bug causes an unexpected behavior.